packages/marshal/src/types.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * @typedef {"bigint" | "boolean" | "null" | "number" | "string" | "symbol" | "undefined" | "copyArray" | "copyRecord" | "copyError" | "promise" | "presence" } PassStyle
+ * TODO "presence" above should indirect through REMOTE_STYLE to prepare
+ * for changing it to "remotable"
+ */
+
+// TODO declare more precise types throughout this file, so the type system
+// and IDE can be more helpful.
+
+/**
+ * @typedef {*} Passable
+ *
+ * A Passable value that may be marshalled. It is classified as one of
+ * PassStyle. A Passable must be hardened.
+ *
+ * A Passable has a pass-by-copy superstructure. This includes the atomic
+ * pass-by-copy primitives ("bigint" | "boolean" | "null" | "number" |
+ * "string" | "undefined") and the composite pass-by-copy objects ("copyArray" |
+ * "copyRecord" | "copyError"). The composite pass-by-copy objects that may
+ * contain other Passables.
+ *
+ * A Passable's pass-by-copy superstructure ends in PassableCap leaves. The
+ * Passable can be further classified by the nature of these leaves. Since a
+ * Passable is hardened, its structure and classification is stable --- its
+ * structure and classification cannot change even if some of the objects are
+ * proxies.
+ */
+
+/**
+ * @typedef {Passable} Comparable
+ *
+ * A Comparable is a Passable in which none of the leaves of the pass-by-copy
+ * superstructure are promises. Two Comparables may be compared by
+ * for equivalence according to `sameStructure`, which is the strongest
+ * equivalence class supported by marshal's distributed object semantics.
+ */
+
+/**
+ * @typedef {Comparable} OnlyData
+ *
+ * A Comparable is OnlyData when its pass-by-copy superstructure has no leaves,
+ * i.e., when all the leaves of the data structure tree are primitive data
+ * types or empty composites.
+ */
+
+/**
+ * @typedef {OnlyData} PureData
+ *
+ * An OnlyData value is PureData when it contains no hidden mutable state,
+ * e.g., when none of its pass-by-copy composite data objects are proxies. This
+ * cannot be determined by inspection. It can only be achieved by trusted
+ * construction. A PureData value cannot be used as a communications channel,
+ * and can therefore be safely shared with subgraphs that should not be able
+ * to communicate with each other.
+ */
+
+/**
+ * @typedef Remotable
```

Comment:
I think `@typedef` needs a type and a name.

```suggestion
 * @typedef {SOMETHING} Remotable
```

Reply:
Done